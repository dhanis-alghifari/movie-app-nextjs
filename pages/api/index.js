import axios from "axios";
import CONFIG from "../config";

const fullURL = (path) => {
  return `${CONFIG.API_URL}/${path}`;
};

export const handleNetworkError = (error) => {
  if (error.message === "Network request failed") {
    alert(
      "Kesalahan Jaringan",
      "Silakan periksa koneksi Anda dan coba kembali.",
      "iconNoInet"
    );
  }
  throw error;
};

const get =
  (api) =>
  async (param = "") => {
    try {
      return await axios(
        `${fullURL(api)}${param}`,
        {
          method: "GET",
          headers: {
            "Content-type": "application/json",
          },
        },
        { handleNetworkError }
      );
    } catch (err) {
      console.log(err);
    }
  };

export const getMovie = get(
  "3/search/movie?api_key=bd73e4973a245701c7e1a2a4a393fddc&language=en-US&page=1&include_adult=false"
);

const API = {
  getMovie,
};

export default API;
