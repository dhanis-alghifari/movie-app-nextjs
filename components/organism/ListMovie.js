import { getMovie } from "../../pages/api";
import { useState, useEffect, useCallback } from "react";
import CONFIG from "../../pages/config";
import Input from "../../components/atom/Input";
import Image from "next/image";

function ListMovie() {
  const [dataMovie, setDataMovie] = useState([]);
  const [search, setSearch] = useState("A");

  const getAllMovie = useCallback(() => {
    getMovie(`&query=${search}`).then((res) => {
      if (res && res.data) {
        var tempList = [];
        tempList = res.data.results;
        setDataMovie(tempList);
      }
    });
  }, [search]);

  useEffect(() => {
    if (search === "") {
      setSearch("A");
    } else {
      getAllMovie();
    }
  }, [search, getAllMovie]);

  const myLoader = ({ src }) => {
    return `${src}`;
  };

  const renderMovieNotFound = () => {
    return (
      <div>
        {(!dataMovie || dataMovie.length === 0) && search !== "A" ? (
          <p>Movie Not Found</p>
        ) : (
          <p>Loading....</p>
        )}
      </div>
    );
  };

  return (
    <>
      <div className="md:flex hidden flex-row flex-wrap items-center lg:ml-auto mr-3 mb-10">
        <div className="m-auto w-2/5">
          <Input onchange={(e) => setSearch(e.target.value)} />
        </div>
      </div>
      <div className="grid grid-cols-3 gap-x-4 gap-y-8">
        {dataMovie?.length > 0
          ? dataMovie?.map((item, index) => (
              <div key={index} className="block">
                <div className="container flex justify-center">
                  <div className="max-w-sm py-10">
                    <div className="bg-white relative shadow-lg hover:shadow-xl transition duration-500 rounded-lg h-[980px]">
                      {item.poster_path === null ? (
                        <Image
                          loader={myLoader}
                          src="https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50?s=400&r=pg"
                          alt="img"
                          width={600}
                          height={800}
                          unoptimized={true}
                        />
                      ) : (
                        <Image
                          loader={myLoader}
                          src={`${CONFIG.Image_URL}${item.poster_path}`}
                          width={600}
                          height={800}
                          alt={item.title ? item.title : "-"}
                          unoptimized={true}
                        />
                      )}
                      <div className="py-6 px-8 rounded-lg bg-white">
                        <h1 className="text-gray-700 font-bold text-2xl mb-3">
                          Title Movie: {item.title ? item.title : "-"}
                        </h1>
                        <p className="text-gray-700 tracking-wide">
                          Release Date:
                          {item.release_date ? item.release_date : "-"}
                        </p>
                        <p className="text-gray-700 tracking-wide">
                          Overview: {item.overview ? item.overview : "-"}
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            ))
          : renderMovieNotFound()}
      </div>
    </>
  );
}

export default ListMovie;
