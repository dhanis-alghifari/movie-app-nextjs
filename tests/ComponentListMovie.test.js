import { render, cleanup, waitFor } from "@testing-library/react";
import ListMovie from "../components/organism/ListMovie";
import { getMovie } from "../pages/api/index";

jest.mock("../pages/api", () => ({
  getMovie: jest.fn(),
}));

describe("ListMovie", () => {
  afterEach(cleanup);

  it("should render data movie when it's fetched successfully", async () => {
    const data = [
      {
        id: 1,
        title: "movie 1",
        release_date: "2022-01-01",
        overview: "This is a movie about ...",
      },
      {
        id: 2,
        title: "movie 2",
        release_date: "2022-02-01",
        overview: "This is a movie about ...",
      },
    ];
    getMovie.mockResolvedValue({ data: { results: data } });

    const { container, getByText } = render(<ListMovie />);

    await waitFor(() => {
      expect(getByText("Title Movie: movie 1")).toBeDefined();
      expect(getByText("Title Movie: movie 2")).toBeDefined();
    });

    expect(container).toMatchSnapshot();
  });

  it("should render loading message when the data movie is being fetched", async () => {
    getMovie.mockResolvedValue(() => new Promise(() => {}));

    const { container, getByText } = render(<ListMovie />);

    await waitFor(() => {
      expect(getByText("Loading....")).toBeDefined();
    });

    expect(container).toMatchSnapshot();
  });

  it("should render movie not found when the data movie is being fetched", async () => {
    const data = [];
    getMovie.mockResolvedValue({ data: { results: data } });

    const { container, queryByText } = render(<ListMovie />);

    await waitFor(() => {
      const element = queryByText("Movie Not Found");
      expect(element).toBeDefined();
    });

    expect(container).toMatchSnapshot();
  });
});
