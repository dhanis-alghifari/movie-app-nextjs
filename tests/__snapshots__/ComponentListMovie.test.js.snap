// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListMovie should render data movie when it's fetched successfully 1`] = `
<div>
  <div
    class="md:flex hidden flex-row flex-wrap items-center lg:ml-auto mr-3 mb-10"
  >
    <div
      class="m-auto w-2/5"
    >
      <div>
        <input
          class="border-0 px-3 py-3 placeholder-black mr-4 text-slate-600 relative bg-orange-300 rounded text-sm shadow outline-none focus:outline-none focus:ring w-full pl-10"
          placeholder="Search..."
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    class="grid grid-cols-3 gap-x-4 gap-y-8"
  >
    <div
      class="block"
    >
      <div
        class="container flex justify-center"
      >
        <div
          class="max-w-sm py-10"
        >
          <div
            class="bg-white relative shadow-lg hover:shadow-xl transition duration-500 rounded-lg h-[980px]"
          >
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjAwIiBoZWlnaHQ9IjgwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                alt="movie 1"
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
            <div
              class="py-6 px-8 rounded-lg bg-white"
            >
              <h1
                class="text-gray-700 font-bold text-2xl mb-3"
              >
                Title Movie: 
                movie 1
              </h1>
              <p
                class="text-gray-700 tracking-wide"
              >
                Release Date:
                2022-01-01
              </p>
              <p
                class="text-gray-700 tracking-wide"
              >
                Overview: 
                This is a movie about ...
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="block"
    >
      <div
        class="container flex justify-center"
      >
        <div
          class="max-w-sm py-10"
        >
          <div
            class="bg-white relative shadow-lg hover:shadow-xl transition duration-500 rounded-lg h-[980px]"
          >
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjAwIiBoZWlnaHQ9IjgwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                alt="movie 2"
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
            <div
              class="py-6 px-8 rounded-lg bg-white"
            >
              <h1
                class="text-gray-700 font-bold text-2xl mb-3"
              >
                Title Movie: 
                movie 2
              </h1>
              <p
                class="text-gray-700 tracking-wide"
              >
                Release Date:
                2022-02-01
              </p>
              <p
                class="text-gray-700 tracking-wide"
              >
                Overview: 
                This is a movie about ...
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ListMovie should render loading message when the data movie is being fetched 1`] = `
<div>
  <div
    class="md:flex hidden flex-row flex-wrap items-center lg:ml-auto mr-3 mb-10"
  >
    <div
      class="m-auto w-2/5"
    >
      <div>
        <input
          class="border-0 px-3 py-3 placeholder-black mr-4 text-slate-600 relative bg-orange-300 rounded text-sm shadow outline-none focus:outline-none focus:ring w-full pl-10"
          placeholder="Search..."
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    class="grid grid-cols-3 gap-x-4 gap-y-8"
  >
    <div>
      <p>
        Loading....
      </p>
    </div>
  </div>
</div>
`;

exports[`ListMovie should render movie not found when the data movie is being fetched 1`] = `
<div>
  <div
    class="md:flex hidden flex-row flex-wrap items-center lg:ml-auto mr-3 mb-10"
  >
    <div
      class="m-auto w-2/5"
    >
      <div>
        <input
          class="border-0 px-3 py-3 placeholder-black mr-4 text-slate-600 relative bg-orange-300 rounded text-sm shadow outline-none focus:outline-none focus:ring w-full pl-10"
          placeholder="Search..."
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    class="grid grid-cols-3 gap-x-4 gap-y-8"
  >
    <div>
      <p>
        Loading....
      </p>
    </div>
  </div>
</div>
`;
